server/coordinator/queue/delay_queue.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package queue
+
+import (
+	"container/heap"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/CeresDB/ceresmeta/server/coordinator/procedure"
+	"github.com/pkg/errors"
+)
+
+type ProcedureScheduleEntry struct {
+	procedure procedure.Procedure
+	runAfter  time.Time
+}
+
+type ProcedureDelayQueue struct {
+	lock      sync.RWMutex
+	heapQueue *HeapPriorityQueue
+	index     map[uint64]procedure.Procedure
+	maxLen    int
+}
+
+type HeapPriorityQueue struct {
+	lock sync.RWMutex
+
+	procedures []*ProcedureScheduleEntry
+}
+
+func (q *HeapPriorityQueue) Len() int {
+	return len(q.procedures)
+}
+
+func (q *HeapPriorityQueue) Less(i, j int) bool {
+	if q.procedures[i].runAfter.Before(q.procedures[j].runAfter) {
+		return true
+	}
```

Comment:
I guess it should the negative value to make the min runAfter lie at the end of the `q.procedures`:
```suggestion
	if q.procedures[j].runAfter.Before(q.procedures[i].runAfter) {
		return true
	}
```

Reply:
The dequeue order of elements is determined by the less method. If `procedures[i] < procedures[j]`, the `smallest` element will be dequeued first. If `procedures[i] > procedures[j]`, the `largest` element will be dequeued first. So the current the order is correct.

Reply:
This order is difficult to understand, I will add some notes separately.